ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/bridge/validate/ConflictCheckWithCommittedImpl.scala

```diff
@@ -149,4 +158,66 @@ private[validate] class ConflictCheckWithCommittedImpl(
     }
   }
 
+  private def sameContractData(
+      actual: (Value.VersionedContractInstance, Timestamp),
+      provided: DisclosedContract,
+  )(implicit loggingContext: LoggingContext): Boolean = {
+    // TODO DPP-1026: Note: Not validating transaction version or agreement text here, because we're making those up
+
+    // TODO DPP-1026: Template id validation can probably be removed once we switch from VersionedContractInstance to Value
+    val actualTemplate = actual._1.unversioned.template
+    val providedTemplate = provided.contractInst.unversioned.template
+
+    // TODO DPP-1026: Make sure to compare the whole argument
+    val actualArgument = actual._1.unversioned.arg
+    val providedArgument = provided.contractInst.unversioned.arg
+
+    if (actualTemplate != providedTemplate) {
+      // TODO DPP-1026: fix the logging context, it should include at least the submission id (to track malicious users).
+      logger.warn(s"Disclosed contract ${provided.contractId.coid} has invalid template id")
+      false
+    } else if (actualArgument != providedArgument) {
+      logger.warn(s"Disclosed contract ${provided.contractId.coid} has invalid argument")
+      false
+    } else if (actual._2 != provided.ledgerEffectiveTime) {
+      logger.warn(s"Disclosed contract ${provided.contractId.coid} has invalid ledgerEffectiveTime")
+      false
+    } else {
+      true
+    }
+  }
+
+  private def validateExplicitDisclosure(
+      disclosedContracts: Set[DisclosedContract],
+      loggingContext: LoggingContext,
+      completionInfo: CompletionInfo,
+  )(implicit
+      contextualizedErrorLogger: ContextualizedErrorLogger
+  ): AsyncValidation[Unit] = {
+    disclosedContracts.foldLeft(Future.successful[Validation[Unit]](Right(()))) {
+      case (f, provided) =>
+        f.flatMap {
+          case Right(_) =>
+            indexService
+              .lookupContractAfterInterpretation(provided.contractId)(loggingContext)
+              .map {
+                case None =>
+                  Left(
+                    // Disclosed contract was archived or never existed
+                    // This has intentionally the same error code as the cases below,
+                    // to make sure malicious users cannot learn about the activeness of a third party contract
+                    // by faking the contract stakeholders to get around Daml authorization rules.
+                    DisclosedContractInvalid(provided.contractId, completionInfo)
+                  )
+                case Some(actual) if !sameContractData(actual, provided)(loggingContext) =>
+                  Left(
+                    // Disclosed contract has a bad payload, most likely submitted by a malicious user
```

Comment:
At the risk of stating something you already know: It’s not necessarily the submitting user that was malicious here. If I’m malicious and  disclose an incorrect contract to you and you use that on a submission you might not have had any malicious intentions even if you trigger this error. Ofc the participant cannot distinguish between the two. This is also a good argument why (on Canton) we really want to produce different errors: That way the user can distinguish if they got malicious input from somewhere else or just an outdated value.